Objects/lazyimportobject.c

```diff
@@ -0,0 +1,212 @@
+        Py_INCREF(level);
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        Py_DECREF(level);
+        return NULL;
+    }
+    m->lz_lazy_import = NULL;
+    Py_INCREF(name);
+    m->lz_name = name;
+    Py_INCREF(globals);
+    m->lz_globals = globals;
+    Py_INCREF(locals);
+    m->lz_locals = locals;
+    Py_INCREF(fromlist);
+    m->lz_fromlist = fromlist;
+    m->lz_level = level;
+    m->lz_resolved = NULL;
+    m->lz_resolving = NULL;
+    PyObject_GC_Track(m);
+    return (PyObject *)m;
+}
+
+PyObject *
+PyLazyImport_NewObject(PyObject *from, PyObject *name)
+{
+    PyLazyImportObject *m;
+    if (!from || !PyLazyImport_CheckExact(from) || !name || !PyUnicode_Check(name)) {
+        PyErr_BadArgument();
+        return NULL;
+    }
+    m = PyObject_GC_New(PyLazyImportObject, &PyLazyImport_Type);
+    if (m == NULL) {
+        return NULL;
+    }
+    PyLazyImportObject *lazy_from = (PyLazyImportObject *)from;
+    if (lazy_from->lz_fromlist != NULL && lazy_from->lz_fromlist != Py_None) {
+        PyObject *fromlist = PyTuple_New(1);
+        if (fromlist == NULL) {
+            return NULL;
+        }
+        Py_INCREF(name);
+        if (PyTuple_SetItem(fromlist, 0, name) < 0) {
+            return NULL;
+        }
+        PyObject *new_from = PyLazyImport_NewModule(
+            lazy_from->lz_name, lazy_from->lz_globals, lazy_from->lz_locals, fromlist, lazy_from->lz_level);
+        Py_DECREF(fromlist);
+        if (new_from == NULL) {
+            return NULL;
+        }
+        m->lz_lazy_import = new_from;
+    } else {
+        Py_INCREF(from);
+        m->lz_lazy_import = from;
+    }
+    Py_INCREF(name);
+    m->lz_name = name;
+    m->lz_globals = NULL;
+    m->lz_locals = NULL;
+    m->lz_fromlist = NULL;
+    m->lz_level = NULL;
+    m->lz_resolved = NULL;
+    m->lz_resolving = NULL;
+    PyObject_GC_Track(m);
+    return (PyObject *)m;
+}
+
+static void
+lazy_import_dealloc(PyLazyImportObject *m)
+{
+    PyObject_GC_UnTrack(m);
+    Py_XDECREF(m->lz_lazy_import);
+    Py_XDECREF(m->lz_name);
+    Py_XDECREF(m->lz_globals);
+    Py_XDECREF(m->lz_locals);
+    Py_XDECREF(m->lz_fromlist);
+    Py_XDECREF(m->lz_level);
+    Py_XDECREF(m->lz_resolved);
+    Py_XDECREF(m->lz_resolving);
+    Py_TYPE(m)->tp_free((PyObject *)m);
+}
+
+static PyObject *
+lazy_import_name(PyLazyImportObject *m)
+{
+    if (m->lz_lazy_import != NULL) {
+        PyObject *name = lazy_import_name((PyLazyImportObject *)m->lz_lazy_import);
+        PyObject *res = PyUnicode_FromFormat("%U.%U", name, m->lz_name);
+        Py_DECREF(name);
+        return res;
+    }
+    if (m->lz_fromlist == NULL ||
+        m->lz_fromlist == Py_None ||
+        !PyObject_IsTrue(m->lz_fromlist)) {
+        Py_ssize_t dot = PyUnicode_FindChar(m->lz_name, '.', 0, PyUnicode_GET_LENGTH(m->lz_name), 1);
+        if (dot >= 0) {
+            return PyUnicode_Substring(m->lz_name, 0, dot);
+        }
+    }
```

Comment:
The reason is because in these cases:

```py
import foo.bar.baz.qux as Q
```

It translates to something like:

```
LOAD_CONST               0 (0)
LOAD_CONST               1 (None)
IMPORT_NAME              0 (foo.bar.baz.qux)
IMPORT_FROM              1 (bar)
SWAP                     2
POP_TOP
IMPORT_FROM              2 (baz)
SWAP                     2
POP_TOP
IMPORT_FROM              3 (qux)
STORE_NAME               4 (Q)
POP_TOP
```

...and we end up with a lazy object that looks like this:

```py
Q = LazyObject(
    from=LazyObject(
        from=LazyObject(
            from=LazyObject(
                name="foo.bar.baz.qux",
                fromlist=None,
            ),
            name="bar",
        ),
        name="baz",
    ),
    name="qux",
)
```

So when the when the full object is extracted here (via `lazy_import_name()`), it first extracts the most nested object (which has a name set to `"foo.bar.baz.qux"`), then it goes to adding the current name on each level, and without doing this stripping, it would end up being `"foo.bar.baz.qux.bar.baz.qux"`.

This directly comes from `import.c`'s `PyImport_ImportModuleLevelObject`, where it also strips the front name before importing it. I'll modify the code so it's more obvious.

`fromlist` is (I believe) more complicated than I would have hoped to.